truncate/table_schema.go

```diff
@@ -43,45 +42,38 @@ type tableSchema struct {
 	referencedBy []string
 }
 
-// indexSchema represents secondary index metadata.
-type indexSchema struct {
-	indexName string
-
-	// Table name on which the index is defined.
-	baseTableName string
-
-	// Table name the index interleaved in. If blank, the index is a global index.
-	parentTableName string
+func (t *tableSchema) isCascadeDeletable() bool {
+	return t.parentOnDeleteAction == deleteActionCascadeDelete
 }
 
-func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables, excludeTables []string) ([]*tableSchema, error) {
+const fetchTableSchemasQuery = `
+WITH FKReferences AS (
+	SELECT CCU.TABLE_NAME AS Referenced, ARRAY_AGG(TC.TABLE_NAME) AS Referencing
+	FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS as TC
+	INNER JOIN INFORMATION_SCHEMA.CONSTRAINT_COLUMN_USAGE AS CCU ON TC.CONSTRAINT_NAME = CCU.CONSTRAINT_NAME
+	WHERE TC.TABLE_CATALOG = '' AND TC.TABLE_SCHEMA = '' AND TC.CONSTRAINT_TYPE = 'FOREIGN KEY' AND CCU.TABLE_CATALOG = '' AND CCU.TABLE_SCHEMA = ''
+	GROUP BY CCU.TABLE_NAME
+)
+SELECT T.TABLE_NAME, T.PARENT_TABLE_NAME, T.ON_DELETE_ACTION, IF(F.Referencing IS NULL, ARRAY<STRING>[], F.Referencing) AS referencedBy
+FROM INFORMATION_SCHEMA.TABLES AS T
+LEFT OUTER JOIN FKReferences AS F ON T.TABLE_NAME = F.Referenced
+WHERE T.TABLE_CATALOG = "" AND T.TABLE_SCHEMA = "" AND T.TABLE_TYPE = "BASE TABLE"
+ORDER BY T.TABLE_NAME ASC
+`
+
+// fetchTableSchemas fetches schema information from spanner database.
+// If targetTables is not empty, it fetches only the specified tables.
+func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables []string) ([]*tableSchema, error) {
```

Comment:
fetchTableSchemas has been changed to concentrate on fetching schema information from Spanner.

Reply:
Sounds good!

If `fetchTableSchemas` focuses on fetching schema from Spanner, how about moving `targetTables` filtering to `filterTableSchemas`?

So that each function will be:

- `fetchTableSchemas`: Fetches the schema information from Spanner
- `filterTableSchemas`: Filters out the unnecessary tables with `targetTables` and `excludedTables`

Reply:
Ah, now I understand your intention! 
I will make the necessary modifications and request your review again in a few days.